### DISCUSSION TITLE
Playtime 2 feedback in TouchOSC

### DISCUSSION THREAD

**Discussion by joeyxsam** [helgoboss/helgobox#1148] [Q&A]
I can set up touchosc to send MIDI/OSC and trigger clips to play/stop by toggling "show/hide MIDI triggers".  What I can't do is get a button that triggers a clip in TouchOSC to stay illuminated while the clip is playing and to not be illuminated when the clip is stopped.  I think this is already possible but not sure...is there a particular type of feedback that I need to send from within realearn to enable this?

Related but different, wondering if it's possible to have the cell's progress position sent as feedback to a fader in touchosc so you could see the progress position from within touchosc?  Or is this already possible some other way? 

And a not critical UI thing:  in realearn, when you add a MIDI source, I'm seeing "CC numbe" instead of "CC number".

**Comment by helgoboss** [comment]
> I can set up touchosc to send MIDI/OSC and trigger clips to play/stop by toggling "show/hide MIDI triggers". What I can't do is get a button that triggers a clip in TouchOSC to stay illuminated while the clip is playing and to not be illuminated when the clip is stopped. I think this is already possible but not sure...is there a particular type of feedback that I need to send from within realearn to enable this?

The target "Playtime: Slot transport action" with action "Trigger" sends special numbers as feedback values depending on the clip state (stopped, playing, scheduled-for-stop, recording, etc.). You can see them as the clip stops/plays in the right Help text field at the bottom. You could react to those special values in TouchOSC. As an alternative, there's text feedback available using text expression "{{ target.slot_state.id }}" which uses more descriptive feedback values such as "playtime.slot_state.stopped". You can receive that using OSC string argument.

> Related but different, wondering if it's possible to have the cell's progress position sent as feedback to a fader in touchosc so you could see the progress position from within touchosc? Or is this already possible some other way?
> 
> And a not critical UI thing: in realearn, when you add a MIDI source, I'm seeing "CC numbe" instead of "CC number".

Target "Playtime: Slot seek"

**Reply by joeyxsam** [reply]
Thanks, Benjamin.  I was able to get these things working based on what you said.  See short gif here which shows touchosc desktop interacting with playtime 2, cell progress meter as well!  https://imgur.com/94DHQVj

I'll try getting the record/stop and overdub/play actions to work with a basic layout in touchosc next.  I know I need to arm the appropriate track for this to work right, but I don't have it all working quite right at the moment.  Thanks for the cool script - planning to purchase a copy soon once I've got it a little more figured out :)

**Reply by FakenameMusic** [reply]
> Thanks, Benjamin. I was able to get these things working based on what you said. See short gif here which shows touchosc desktop interacting with playtime 2, cell progress meter as well! https://imgur.com/94DHQVj
> 
> I'll try getting the record/stop and overdub/play actions to work with a basic layout in touchosc next. I know I need to arm the appropriate track for this to work right, but I don't have it all working quite right at the moment. Thanks for the cool script - planning to purchase a copy soon once I've got it a little more figured out :)

Awesome work! Touch OSC is something I'd really love to use with Playtime 2. I tried but fell into the same issue that you had that you resolved here. 🤘

**Reply by joeyxsam** [reply]
Hi Benjamin,

In trying to setup a basic template in touchosc to work with playtime 2, I feel I may be trying to reinvent the wheel and not effectively utilizing what you've already built.  With this in mind, I'm wondering if you already have a touchosc template for playtime 2?  Or if it would be possible to translate one of your existing presets, like the apc key 25, to a touchosc layout.  If the various buttons on the apc key 25 are sending/receiving midi from realearn, then I could probably make a similar layout in touchosc, but I have no idea what the apc key 25 is sending/receiving.  In that preset, I see the source is "virtual" for every action, but not sure what that means or why I would want virtual source vs OSC/MIDI.  Thanks for any insight!

**Reply by helgoboss** [reply]
I don't have a TouchOSC template, neither a ReaLearn preset for it. I'm personally not so interested in touch control (except when it comes to the Playtime App Android/iOS version). But I think it should be relatively straightforward to make a template and preset. I don't think that the APC Key 25 preset is the best place to look at. It contains a lot of complex logic that you probably don't need.

As always with ReaLearn, the best way is to start small and then grow bigger. Create a super simple template with just one button and map it to the "Playtime: Slot transport action" target with action "Trigger". The "control" direction should be super simple and become a quick success (well, if you know the basics of ReaLearn, but the video tutorials cover this). Feedback is also not that complicated. You would need a separate feedback-only mapping with an OSC source of argument type String, I guess. Then you switch the feedback to "Text feedback" in the Glue section. This makes the target send quite straightforward strings (e.g. "playtime.slot.playing") to indicate the current slot status. You can visualize this in TouchOSC accordingly.

You can use real sources instead of virtual ones. For OSC, virtual sources don't make that much sense. The sources are already named (they have a usually self-explaining OSC address) and TouchOSC templates are usually very custom, which makes the main benefit of virtual control elements - reusability - not relevant here.

Or you can wait until I do this myself one day. But don't know when that will happen because there are much more urgent topics for me.

**Comment by RibbleFripFnibbulator** [comment]
Thanks for the info all, I'm able to see "slot seek" coming into TouchOSC, showing current seconds. Is there a way to get the total seconds? Better yet, is there a list of all the feedback properties we can target somewhere?

**Reply by RibbleFripFnibbulator** [reply]
I discovered that changing the Slot Seek mapping from Textual feedback to Numeric feedback will return the percentage progress instead of seconds. This will save me some time from fudging the total clip lengths from max seconds values 😁  Onward!
